### PR TITLE
Refactor pipeline execution functionality to allow for multiple backends

### PR DESCRIPTION
**Pull request by emiehling**
- Add aisteer360/backends/ with a common backend interface to support (in-process) Hugging Face backend (backends/huggingface/) and a vLLM backend (backends/vllm/, offline engine and server); vLLM server functionality supported by merged [vLLM-Hook changes](https://github.com/IBM/vLLM-Hook/tree/steerability-challenge) 
- Implement single backend per pipeline and a deterministic acquire/release lifecycle with compatibility check before load/inference
- Add offline fitting and capture sessions so activation methods are fit through the backend 
- Introduce an intervention representation for state controls
- Add the constrained_decoding output control (automaton-based)
- Unify the evaluation generation path, make the Benchmark backend-aware, and route judge metrics and Perplexity through the backend (evaluation/metrics/backend_utils.py)
- RAD / SASA fixes
- Support for reasoning models
- Restructure gating into structured gates and relocate routing / extract pipeline internals into modules
- Update docs, README, notebooks
- Rerun notebooks
- Add/modify tests for backend, session lifecycle, vLLM engine/plugin/release/serve, intervention IR/lowering/export, constrained-decoding, RAD, verbosity, and backend-execution 
- Bump version to 0.4.0
- Closes #17

### CODE REVIEW

File: aisteer360/backends/huggingface/session.py [hunk 1]
```diff
@@ -0,0 +1,697 @@
...
+        params: GenerationParams,
+        gen_kwargs: dict,
+        user_processors: tuple,
+        user_criteria: tuple,
+        seed: int | None,
+    ) -> list[ItemResult]:
+        """One `model.generate` pass over all items (identical entries, one shared seed)."""
+        model = self.model
+        rows = [self._resolve_prompt_tensors(item.prompt) for item in items]
+        input_ids, attention_mask = self._stack_prompt_rows(rows)
```
Comment (ingelise) [inline aisteer360/backends/huggingface/session.py:443]: @emiehling, There is an issue with padding here where batched generation right-pads prompts on the HF backend, which you can solve the same way you have in _score_batched by adding
`input_ids, attention_mask = to_left_pad(input_ids, attention_mask)   `

File: aisteer360/backends/vllm/backend.py [hunk 2]
```diff
@@ -0,0 +1,607 @@
...
+        model_ref = checkpoint.path if checkpoint is not None else spec.model
+        if model_ref is None:
+            raise ValueError("VLLMBackend needs a model reference on the spec or a checkpoint artifact.")
+        trust_remote_code = bool(spec.get_option("trust_remote_code", default=False))
+        _reject_encoder_decoder(model_ref, trust_remote_code)
+
+        engine_kwargs = dict(spec.get_option("engine_kwargs", default={}) or {})
+        # default to a compact grammar so json constraints match the in-process automaton
+        # (disable_any_whitespace needs an explicit backend); caller kwargs win
+        engine_kwargs.setdefault("structured_outputs_config", {"disable_any_whitespace": True, "backend": "xgrammar"})
```
Comment (ingelise) [inline aisteer360/backends/vllm/backend.py:175]: This can fail in older vllm .  Your Pyproject toml has vllm>=0.8.5,<1.0.0 .  Maybe you can increase it there?

File: aisteer360/evaluation/metrics/backend_utils.py [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+guidance in the `Backend` docstring.
+"""
+from __future__ import annotations
+
+from aisteer360.algorithms.core.execution.backend import Backend, resolve_backend_class
+from aisteer360.algorithms.core.execution.spec import BackendSpec
+
+BackendConfig = "BackendSpec | str | Backend | None"
+
+_METRIC_BACKENDS: dict[BackendSpec, Backend] = {}
```
Comment (ingelise) [inline aisteer360/evaluation/metrics/backend_utils.py:16]: One of the AI assisted code review suggestions: 
```
3. Metric backends are cached process-globally with no release path

  evaluation/metrics/backend_utils.py:16 — _METRIC_BACKENDS is a module-level dict with no eviction and no public release API. Two consequences:

  - A judge or Perplexity configured with backend="vllm" boots an engine that is never shut down. Benchmark correctly calls pipeline.release_backends() between configs (benchmark.py:514), but
  nothing reaches the metric cache.
  - Because _METRIC_BACKENDS and SteeringPipeline._backends are separate dicts, a vLLM judge alongside a vLLM pipeline yields two live engines in one process — which VLLMBackend.release()
  explicitly documents as unsupported: "The distributed-state teardown is process-global, so release assumes no other live vLLM engine in the process." Releasing the pipeline's engine calls
  destroy_model_parallel() / destroy_distributed_environment() and pulls the floor out from under the still-cached judge engine.

  The tests already reach into the private dict (backend_utils._METRIC_BACKENDS.clear() in test_perplexity.py:188, test_base_judge.py:247), which is a good signal the public surface is missing.
  A release_metric_backends() plus a documented note on the single-engine constraint would close it.```

File: tests/internals/test_fingerprint.py [hunk 4]
```diff
@@ -63,3 +63,26 @@ def test_config_changes_digest(self, saved_model_dir):
         torch.manual_seed(0)
         b = tiny_llama(num_layers=2)
         assert model_fingerprint(a) != model_fingerprint(b)
+
```
Comment (ingelise) [inline tests/internals/test_fingerprint.py:66]: Should this file and others use vllm_hook_plugins importorskip?
- tests/internals/test_fingerprint.py
- tests/core/test_spec_hook_equivalence.py      (collection error)
- tests/core/test_vllm_serve_backend.py         (8 failures)
- tests/core/test_steering_pipeline.py
- tests/core/test_intervention_lowering.py      (collection error)
- tests/controls/test_intervention_export.py    (collection error)
- tests/core/test_declarative_phases.py
- tests/core/test_driver_rollout_anchor.py
- tests/controls/test_intervention_ir.py

File: aisteer360/backends/vllm/backend.py [hunk 5]
```diff
@@ -0,0 +1,607 @@
...
+            # explicit False, so this only fills the default
+            engine_kwargs.setdefault("enforce_eager", True)
+
+        # the worker-selection variable is scoped to this engine's boot so a later plugin-free
+        # engine in the same process is unaffected
+        previous_worker = os.environ.get("VLLM_HOOK_WORKER")
+        if spec.get_option("hook_plugin"):
+            os.environ["VLLM_HOOK_WORKER"] = "unified"
+        try:
+            self._llm = LLM(model=model_ref, **engine_kwargs)
```
Comment (ingelise) [inline aisteer360/backends/vllm/backend.py:191]: VLLMBackend.__init__ boots the engine here, then resolves tokenizer (208), layout (209), and discovery (214). If any of those  have errors the engine will be live but the instance never return,
  therefore  _backend_for never records it in self._backends and steer()'s except: release_backends() cannot reach it. 

You could wrap everything after LLM(...) in a try/except that calls self.release() before re-raising.

File: aisteer360/backends/vllm/backend.py [hunk 6]
```diff
@@ -0,0 +1,607 @@
...
+                expect_json=False,
+            )
+        except (TransportError, ValueError) as error:
+            raise ValueError(
+                f"Could not load the LoRA artifact at {lora.path!r} onto the server "
+                f"(dynamic adapter loading requires VLLM_ALLOW_RUNTIME_LORA_UPDATING): {error}"
+            ) from error
+        return adapter_name
+
+    def _verify_fingerprints(self, tokenizer_source: str) -> None:
```
Comment (ingelise) [inline aisteer360/backends/vllm/backend.py:527]: tokenizer_source is not used here.

File: aisteer360/backends/vllm/backend.py [hunk 7]
```diff
@@ -0,0 +1,607 @@
...
+    constructing the backend.
+    """
+
+    def __init__(self, spec: BackendSpec, artifacts: Sequence[Artifact] = ()) -> None:
+        if spec.kind != "vllm":
+            raise ValueError(f"VLLMBackend requires a 'vllm' spec; got kind {spec.kind!r}.")
+        self.spec = spec
+        self._released = False
+        require("vllm")
+        import os
```
Comment (ingelise) [inline aisteer360/backends/vllm/backend.py:161]: This is imported at top of file already

**Review by ingelise — CHANGES_REQUESTED** [review]
Hi @emiehling, 

I have some notes on code as comments here, mainly distilled from the AI code review.  

I had a few test failures on CCC - logs mentioned to you earlier.

For the AI assisted code review:
## Reproduction environment
```
python 3.11 · venv at /workspace/.venv-rev (ephemeral)
torch 2.13.0+cpu · transformers 4.56.2 (inside the declared >=4.52,<5.0 range)
scikit-learn, pandas, pytest installed; vllm / vllm_hook_plugins / xgrammar NOT installed
trl 1.10.0 and peft 0.20.0 were OUT of range (pins are trl<0.28, peft<0.16) —
  this accounts for 3 of the 18 suite failures and nothing else
```

Suite command used:

```bash
python -m pytest tests/ -q --no-header -p no:cacheprovider \
  --ignore=tests/controls/test_ppo_wrapper.py --ignore=tests/controls/test_trl_release.py \
  --ignore=tests/core/test_registry.py --ignore=tests/core/test_spec_hook_equivalence.py \
  --ignore=tests/controls/test_intervention_export.py --ignore=tests/core/test_intervention_lowering.py
# -> 18 failed, 2179 passed, 727 skipped in 298s
```

You might consider documenting the breaking changes somewhere as well.  For example  common to _common.

I will continue running through the notebooks and will flag if I come across any errors.

File: aisteer360/algorithms/core/execution/params.py [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+            stop_strings = (stop_strings,)
+        merged = list(params.stop_strings)
+        merged.extend(text for text in stop_strings if text not in merged)
+        updates["stop_strings"] = tuple(merged)
+    stop_token_ids = contribution.get("stop_token_ids")
+    if stop_token_ids:
+        merged_ids = list(params.stop_token_ids)
+        merged_ids.extend(int(i) for i in stop_token_ids if int(i) not in merged_ids)
+        updates["stop_token_ids"] = tuple(merged_ids)
+    max_new = contribution.get("max_new_tokens")
```
Comment (ingelise) [inline aisteer360/algorithms/core/execution/params.py:192]: Question, do you need some checks or logic if min can be bigger than max tokens?

File: aisteer360/backends/vllm/backend.py (hunk 2 above)
Comment (ingelise) [reply aisteer360/backends/vllm/backend.py]: Hi, this is still an issue, recommend you increase the vllm version in pyproject toml

File: aisteer360/evaluation/metrics/backend_utils.py (hunk 3 above)
Comment (ingelise) [reply aisteer360/evaluation/metrics/backend_utils.py:25]: Hi I think this still needs to be addressed.

**Review by ingelise — CHANGES_REQUESTED** [review]
Hi @emiehling thanks for the changes.  Just one or two more that could  still be addressed.

File: aisteer360/algorithms/core/execution/params.py (hunk 8 above)
Comment (emiehling) [reply aisteer360/algorithms/core/execution/params.py:192]: Added a check

File: aisteer360/evaluation/metrics/backend_utils.py (hunk 3 above)
Comment (emiehling) [reply aisteer360/evaluation/metrics/backend_utils.py:25]: Now releasing the backends

File: aisteer360/backends/vllm/backend.py (hunk 2 above)
Comment (ingelise) [reply aisteer360/backends/vllm/backend.py]: Hi, @emiehling I recommend you increase the vllm version in pyproject.toml file to  >=0.24.0,<1.0.0 so you can avoid this error and any CVEs

File: pyproject.toml [hunk 9]
```diff
@@ -65,6 +65,13 @@ plots = [
     "matplotlib>=3.8.0,<4.0.0",
     "seaborn>=0.13.2,<0.14.0",
 ]
+guided = [
+    "xgrammar>=0.1.19",
+]
+vllm = [
+    "vllm>=0.8.5,<1.0.0",
```
Comment (ingelise) [inline pyproject.toml:72]: consider changing to this : `>=0.24.0,<1.0.0`

File: aisteer360/evaluation/metrics/backend_utils.py (hunk 3 above)
Comment (ingelise) [reply aisteer360/evaluation/metrics/backend_utils.py:25]: Hi I see you have added a function for this but I can't find where it is called?

**Review by ingelise — CHANGES_REQUESTED** [review]

File: aisteer360/evaluation/metrics/backend_utils.py (hunk 3 above)
Comment (emiehling) [reply aisteer360/evaluation/metrics/backend_utils.py:25]: I've added a call in `Benchmark.run()` to release when the run finishes or fails

File: aisteer360/backends/vllm/backend.py (hunk 2 above)
Comment (emiehling) [reply aisteer360/backends/vllm/backend.py]: We're going to be bumping vllm up in the refactor of the evaluation functionality (it's coupled with some other changes that we're making that use newer versions of `trl`, `transformers`, etc.)

File: aisteer360/evaluation/metrics/backend_utils.py (hunk 3 above)
Comment (ingelise) [reply aisteer360/evaluation/metrics/backend_utils.py:25]: It looks like you have added the function `release_metric_backends `but it is never called.  A suggestion would be to add a call to it after pipeline.release_backends() in benchmark.py:514

**Review by ingelise — APPROVED** [review]
